### PR TITLE
feat: [engine] enforce business id uniqueness per process definition

### LINKED ISSUE #45179: Enforce business key uniqueness per process definition on instance start
### 

### Description

Enforce that a root process instance can only be started with a given business key if no other *running or existing* root instance of the same process definition already uses that business key.

Scope and rules (from parent issue #44978):
- Uniqueness scope: per process definition
  - per-tenant if multi-tenancy is enabled
  - otherwise per cluster
- Only root process instances (children/call activities out of scope for now)
- Only applies to instances created after enforcement is ON (no retroactive enforcement)
- Must implement required partitioning/routing semantics so the check is deterministic
- Must fail the start request with a deterministic error contract when violated

### Motivation

Deterministic idempotency is a top customer requirement for the 8.9 MVP. Enforcing business key uniqueness per process definition closes a critical migration gap vs Camunda 7 and matches customer expectations for safe retries of process start without duplicates.

### Acceptance criteria

- [ ] Starting a root process instance with a business key already in use for the same process definition is rejected
- [ ] Rejection uses a deterministic, documented error contract
- [ ] Uniqueness scope is per process definition and per-tenant when MT is enabled (otherwise per cluster)
- [ ] Child/call-activity instances are not subject to this check
- [ ] Enforcement applies only to process instances created after enforcement is enabled
- [ ] Partitioning/routing semantics ensure deterministic behavior across partitions/nodes
- [ ] Regression suite includes negative tests covering duplicate starts and relevant edge cases
- [ ] Design does not block a future configuration toggle (see parent issue)

### Links

Parent issue: https://github.com/camunda/camunda/issues/44978
Context / edge cases:
https://github.com/camunda/product-hub/issues/2003#issuecomment-3755709033
https://github.com/camunda/product-hub/issues/2003#issuecomment-3759886403
https://github.com/camunda/product-hub/issues/2003#issuecomment-3778233243
https://camunda.slack.com/archives/C0A63CM61NU/p1769604844078599

### Breakdown

- Persist the the process instance when business id is set in a way that allows easy lookup by process definition key, tenant id, and business id to check that one already exists
- Check that one exists on instance creation and return error appropriately
- Implement routing/partitioning strategy to ensure the check and write are consistent
- Add deterministic error response (type/message/metadata) for uniqueness violations
- Add regression tests:
  - duplicate start (same process def + business key)
  - different process defs with same business key (allowed)
  - MT: same tenant vs different tenant behavior
  - non-root instances unaffected
  - behavior when enforcement is toggled on (no retroactive enforcement)

### PR DESCRIPTION
**Pull request by mustafadagher**
## Description

This PR enforces **business id uniqueness per process definition (scoped by tenant)** when creating process instances.

When a `businessId` is provided on `ProcessInstanceCreation`, the engine now checks for an existing process instance with the same `(businessId, processDefinitionKey, tenantId)` and rejects the command with `ALREADY_EXISTS` if one is found.

To support this efficiently, it introduces a new state column family (`PROCESS_INSTANCE_KEY_BY_BUSINESS_ID`) that maps `(businessId, processDefinitionKey, tenantId)` to a process instance key. The index is populated on process instance activation (new `ELEMENT_ACTIVATING` applier v3). Tests cover the rejection behavior and tenant scoping (same business id is allowed across tenants).

## Checklist

- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://camunda.github.io/camunda/ci/#when-to-backport-ci-changes)).

## Related issues

closes #45179

### CODE REVIEW

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbElementInstanceState.java [hunk 1]
```diff
@@ -77,6 +77,16 @@ public final class DbElementInstanceState implements MutableElementInstanceState
...
 
+  // Business ID indexing: [businessId | processDefinitionKey | tenantId] => processInstanceKey
+  private final DbString businessId = new DbString();
+  private final DbString tenantId = new DbString();
+  private final DbCompositeKey<DbString, DbLong> businessIdAndProcessDefinitionKey;
+  private final DbCompositeKey<DbCompositeKey<DbString, DbLong>, DbString>
+      businessIdProcessDefinitionAndTenantKey;
+  private final DbLong processInstanceKeyByBusinessId = new DbLong();
+  private final ColumnFamily<DbCompositeKey<DbCompositeKey<DbString, DbLong>, DbString>, DbLong>
+      rootProcessInstanceKeyByBusinessIdColumnFamily;
```
Comment (korthout) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbElementInstanceState.java]: ❌ This column family structure will not allow us to have multiple instances of the same process definition for the same tenant with the same businessId. This is a requirement when idempotency is disabled.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementActivatingV3Applier.java
Comment (korthout) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementActivatingV3Applier.java:1]: We should introduce a new applier in separate commits:
1. Make a copy of the latest version `x`, rename it to `Vx+1`
2. Make the changes in `Vx+1`

This means we leave the old version alone, and we introduce the new version in just two commits where one shows clearly the changes. I believe it's the easiest to read for future git archeologists.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementActivatingV2Applier.java [hunk 2]
```diff
@@ -140,6 +140,9 @@ private void applyRootProcessState(
         elementInstance.setProcessDepth(parentProcessInstance.getProcessDepth() + 1);
         elementInstanceState.updateInstance(elementInstance);
       }
+    } else {
+      // insert business id index for root process instances
+      insertBusinessIdIndex(value);
```
Comment (korthout) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementActivatingV2Applier.java]: ❌ It seems to me like we'd only do it for root process instances, but what about child process instances?

🔧 We can do it simpler. If the value is set in the record, then we need to store it in the column family. We can do that somewhere outside of this `applyRootProcessState` method.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceCreationHelper.java [hunk 3]
```diff
@@ -412,6 +423,29 @@ private boolean hasAdHocSubProcessScope(final ExecutableFlowElement flowElement)
...
+      final String businessId, final long processDefinitionKey, final String tenantId) {
+    // If no business id is provided, skip validation
+    if (businessId == null || businessId.isEmpty()) {
+      return VALID;
+    }
+
+    // Check if a process instance with this business id already exists
+    final long existingProcessInstanceKey =
+        elementInstanceState.getRootProcessInstanceKeyByBusinessId(
+            businessId, processDefinitionKey, tenantId);
```
Comment (korthout) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceCreationHelper.java]: :x: We shouldn't do this by root process instance but by process instance.

Consider the following case:
- Process A has a call activity that can call B
- Process B has some task
- I can create an instance of A with business id 123
- I can create an instance of B with business id 234
- I can create an instance of A with business id 234, it should raise an incident  at the call activity when it tries to create an instance of B with business id 234 as there's already one running

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceCreationHelper.java [hunk 4]
```diff
@@ -412,6 +423,29 @@ private boolean hasAdHocSubProcessScope(final ExecutableFlowElement flowElement)
...
+
+    if (existingProcessInstanceKey != -1) {
```
Comment (korthout) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceCreationHelper.java]: 💡 What we need here is an `.exists` check, not a full read

Currently the DB doesn't support `.existsWithPrefix` but that's something we could consider adding.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceCreationHelper.java [hunk 5]
```diff
@@ -195,6 +202,10 @@ public Either<Rejection, DeployedProcess> validateCommand(
         .flatMap(
             valid -> validateElementNotBelongingToEventBasedGateway(process, startInstructions))
         .flatMap(valid -> validateTags(tags))
+        .flatMap(
+            valid ->
+                validateBusinessIdUniqueness(
+                    businessId, deployedProcess.getKey(), command.getTenantId()))
```
[the remark this answers is not in the document]
Comment (korthout) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceCreationHelper.java:208]: Correct. However, we'll add this in a later PR. See https://github.com/camunda/camunda/issues/44979

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/immutable/ElementInstanceState.java [hunk 6]
```diff
@@ -133,6 +133,18 @@ Set<DirectBuffer> getTakenSequenceFlows(
...
       long processInstanceKey, String elementId);
 
+  /**
+   * Returns the process instance key for the given business id, process definition key, and tenant
+   * id. Returns -1 if no active process instance exists with the given business id.
+   *
+   * @param businessId the business id to look up
+   * @param processDefinitionKey the process definition key
+   * @param tenantId the tenant id
+   * @return the process instance key, or -1 if not found
```
[the remark this answers is not in the document]
Comment (korthout) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/immutable/ElementInstanceState.java]: >or implement deletion of the index entry when an instance should no longer reserve the businessId.

👍 Good point. We need to add clean up still.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbElementInstanceState.java [hunk 7]
```diff
@@ -512,6 +548,28 @@ public List<RuntimeInstructionValue> getRuntimeInstructionsForElementId(
...
+
+    final AtomicLong processInstanceKey = new AtomicLong(-1);
+    processInstanceByBusinessIdIndexKeyColumnFamily.whileEqualPrefix(
+        businessIdIndexKey,
+        (key, value) -> {
+          processInstanceKey.set(key.processInstanceKey());
+          // just find the first, and then stop iterating
+          return false;
+        });
+    return processInstanceKey.get();
```
[the remark this answers is not in the document]
Comment (korthout) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbElementInstanceState.java]: 👍 Good suggestion

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceTest.java [hunk 8]
```diff
@@ -570,4 +568,94 @@ public void shouldActivateNoneStartEventWhenEventTriggerIsAvailableInState() {
             tuple(
                 BpmnElementType.START_EVENT, "noneStart", ProcessInstanceIntent.ELEMENT_ACTIVATED));
   }
+
+  @Test
+  public void shouldRejectCreateProcessInstanceWithBusinessIdInUse() {
+    final String processId = helper.getBpmnProcessId();
+    final String businessId = "biz-123";
+
```
[the remark this answers is not in the document]
Comment (korthout) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceTest.java:576]: 👍 Makes sense, but will come later. See #44979

File: zeebe/protocol/src/main/java/io/camunda/zeebe/protocol/ZbColumnFamilies.java [hunk 9]
```diff
@@ -266,7 +266,9 @@ public enum ZbColumnFamilies implements EnumValue, ScopedColumnFamily {
   JOB_METRICS_STRING_ENCODING(135, PARTITION_LOCAL),
   JOB_METRICS_META(136, PARTITION_LOCAL),
 
-  CONDITIONAL_SUBSCRIPTION_BY_TENANT_ID(137, PARTITION_LOCAL);
+  CONDITIONAL_SUBSCRIPTION_BY_TENANT_ID(137, PARTITION_LOCAL),
+
+  PROCESS_INSTANCE_KEY_BY_BUSINESS_ID(138, PARTITION_LOCAL);
```
[the remark this answers is not in the document]
Comment (korthout) [reply zeebe/protocol/src/main/java/io/camunda/zeebe/protocol/ZbColumnFamilies.java:271]: 👎 Not going to change this to `GLOBAL`. Instead, we'll dispatch to specific partitions with deterministic routing. See #44979

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbElementInstanceState.java (hunk 1 above)
Comment (mustafadagher) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbElementInstanceState.java]: We agreed in discussion that both options are possible ways, and we will open a discussion with PM to choose which option to go for:
Whether to keep track of business id for all running processes, or only to do so if uniqueness is enabled.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementActivatingV2Applier.java (hunk 2 above)
Comment (mustafadagher) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementActivatingV2Applier.java]: We discussed this, it's an interesting use case where we might have instances from the same process definition: ones that were created by a call activity inheriting a business key, and another that might be created as root using the same (inherited before) business key.

I agree with you that it might be wise to disable such pattern. Let's just documented and make it clear. Maybe also worth a message on the project channel to notify the PM about such case

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbElementInstanceState.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (mustafadagher) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbElementInstanceState.java]: Maybe it's just better to only return a boolean (true/false) from this method. Since actually, now, the returned key is just one of many

[the remark this answers is not in the document]
Comment (korthout) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbElementInstanceState.java]: I've added 064ac31. Please have another look 🙇

**Review by korthout — APPROVED** [review]
LGTM 👍
